File: 2025/day-02/part2.py
import sys

if len(sys.argv) > 1:
	INPUT = sys.argv[1]
else:
	INPUT = "input2.txt"

invalid_ids = 0
with open(INPUT) as f:
	ranges = f.read().strip().split(",")
	for cur_range in ranges:
		rangesplit = cur_range.split("-")
		for i_outer in range(int(rangesplit[0]), int(rangesplit[1]) + 1):
			numstr = str(i_outer)
			mid = int(len(numstr) / 2)
			left, right = numstr[:mid], numstr[mid:]
			if left == right:
				invalid_ids += i_outer
				print(i_outer)
			else:
				first_seq = numstr[0]
				first_rep = None
				# last_seq = None
				maybe_seq = None
				for i in range(1, len(numstr)):
					if numstr[i] == first_seq:
						# print("repeated number: ", left[i])
						# there might be a sequence
						first_rep = i
						# last_seq = i - 1
						maybe_seq = left[0:first_rep]
						# print(maybe_seq)
						if len(numstr) % len(maybe_seq) == 0:
							needed_reps = int(len(numstr) / len(maybe_seq))
							# print(f"sequence {maybe_seq} needs to repeat {needed_reps} to equal {numstr}")
							if maybe_seq * needed_reps == numstr:
								print(f"found {maybe_seq} in {numstr}")
								invalid_ids += i_outer
							# else:
							# 	print(f"{maybe_seq} not repeating correctly for {numstr}")
						break

print(invalid_ids)

File: 2025/day-02/src/part2.rs
use itertools::Itertools;
use miette::IntoDiagnostic;

#[tracing::instrument]
pub fn process(_input: &str) -> miette::Result<String> {
	let mut invalid_ids = 0;
	for range in _input.split(",") {
		let rangesplit: (&str, &str) = range
			.trim()
			.split("-")
			.collect_tuple()
			.expect("two elements");
		let (left, right) = (
			rangesplit.0.parse::<u64>().into_diagnostic()?,
			rangesplit.1.parse::<u64>().into_diagnostic()?,
		);
		for i in left..=right {
			let numstr = i.to_string();
			let mid: usize = numstr.len() / 2;
			if numstr[..mid] == numstr[mid..] {
				invalid_ids += i;
			} else {
				let first = numstr.chars().next().expect("first");
				for (ii, c) in numstr[1..].chars().enumerate() {
					if c == first {
						let maybe_seq = &numstr[..=ii];
						if numstr.len().checked_rem(maybe_seq.len()) == Some(0)
							&& maybe_seq.repeat(numstr.len() / maybe_seq.len()) == numstr
						{
							// println!("found {maybe_seq} in {numstr}");
							invalid_ids += i;
						}
						break;
					}
				}
			}
		}
	}
	Ok(invalid_ids.to_string())
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn test_process() -> miette::Result<()> {
		let input = "11-22,95-115,998-1012,1188511880-1188511890,222220-222224,1698522-1698528,446443-446449,38593856-38593862,565653-565659,824824821-824824827,2121212118-2121212124";
		assert_eq!("4174379265", process(input)?);
		Ok(())
	}
}
